src/main/java/io/zeebe/kafka/connect/source/ZeebeSourceTask.java

```diff
@@ -68,39 +68,38 @@ public void start(final Map<String, String> props) {
     workers =
         jobTypes.stream().map(type -> this.newWorker(config, type)).collect(Collectors.toList());
 
-    running.set(true);
+    if (!taskState.compareAndSet(TaskState.STOPPED, TaskState.STARTED)) {
+      throw new IllegalArgumentException(
+          String.format(
+              "Expected task to be stopped before start() is called, but current state is '%s'",
+              taskState.get()));
+    }
   }
 
+  @SuppressWarnings("squid:S1168")
   @Override
   public List<SourceRecord> poll() throws InterruptedException {
-    if (running.get()) {
-      final List<ActivatedJob> activatedJobs = new ArrayList<>();
-
-      // if no jobs available, block a few seconds until we receive one
-      if (jobs.isEmpty()) {
-        LOGGER.trace("No jobs available, block for {}ms", JOB_QUEUE_TIMEOUT_MS);
-        final ActivatedJob job = jobs.poll(JOB_QUEUE_TIMEOUT_MS, TimeUnit.MILLISECONDS);
-        if (job == null) {
-          LOGGER.trace("No jobs available, returning control to caller");
-          return null;
-        }
+    final List<SourceRecord> records = new ArrayList<>();
+    if (taskState.compareAndSet(TaskState.STARTED, TaskState.POLLING)) {
+      drainQueueBlocking().stream().map(this::transformJob).forEach(records::add);
+      taskState.compareAndSet(TaskState.POLLING, TaskState.STARTED);
```

Comment:
What if this fails?

Reply:
But if stop is called it will not change the POLLING state right? How we clean up if we are in the polling phase :sweat_smile: